### DISCUSSION TITLE
Tabbing out of osu!lazer takes excessively long

### DISCUSSION THREAD

**Discussion by Joppe27** [ppy/osu#16914] [Q&A]
Hi there. I'm the one who complained about SDL in the last community meeting, and finally got around to recording one of the issues that is definitely the worst of the few that are left. I will already mention right now that all of my drivers are up-to-date, discord overlay is off, and that nothing about my setup is super unusual (laptop does not have a discrete gpu or optimus, only Intel HD). This is on Windows 11.

<details>
  <summary>Video and explanation</summary>
I have no idea what the game is doing here exactly, but when I tab out it starts blackscreening my laptop for a few times only to come back as if I hadn't tabbed out at all. I need to to tab out again for it to actually let me go to another window (after more blackscreening). This entire process takes about half a minute or so which is not fun:

https://user-images.githubusercontent.com/56885706/154774357-dac6b76d-5af0-4eee-93fa-2a31fbe52634.mp4
</details>

What makes me think that the problem isn't at my end is that other apps that work just fine, and that this pc has no (gpu) issues in general. Here are some very short videos of other apps (including stable):

<details>
  <summary>osu! stable</summary>

https://user-images.githubusercontent.com/56885706/154775268-52623f78-2eca-4d12-af26-d13401558a54.mp4
</details>

<details>
  <summary>League</summary>

https://user-images.githubusercontent.com/56885706/154775404-34f3a900-2c62-43d5-8d11-1d233a449932.mp4
</details>

<details>
  <summary>Simcity 2013 (notorious for being a buggy mess but does pretty okay here)</summary>

https://user-images.githubusercontent.com/56885706/154775566-9219ca5a-1620-4822-87b3-c1ddffb98b0f.mp4
</details>

<details>
  <summary>Spin Rhythm</summary>

https://user-images.githubusercontent.com/56885706/154775519-280b574c-16ac-441d-8b1f-be92c6555bb8.mp4
</details>

It feels like osu!lazer's fullscreen is too intrusive(?)/at a deep level if it can turn of my entire display like that, but that's just guessing.

Logs for specs:
[logs.zip](https://github.com/ppy/osu/files/8101064/logs.zip)

**Comment by frenzibyte** [comment]
I believe you're running osu!lazer in exclusive fullscreen mode, not borderless, right? The "Render at native resolution" option in osu!stable would make it run in borderless afaik, and it looks that you're having that enabled:
![CleanShot 2022-02-19 at 06 19 20](https://user-images.githubusercontent.com/22781491/154784126-8c70d2e9-c112-49cf-be0a-5b8acb018bf1.png)

You will need to disable that to enter exclusive fullscreen mode in osu!stable, for accurate comparsion with osu!lazer.

That said, the tabbing delay in osu!lazer does look excessively long anyways, would probably require further investigation and potentially a report to upstream if not an issue with our window implementation (that is, if reproducible on other machines and not potentially a bug in Windows 11).

**Reply by peppy** [reply]
Yep, it does sound like the user is running in exclusive fullscreen but wants to be running in borderless window. Note that this will increase input latency as a result.

**Reply by Joppe27** [reply]
> You will need to disable that to enter exclusive fullscreen mode in osu!stable, for accurate comparsion with osu!lazer.

No, "render at native resolution" is the toggle for letterboxing and has nothing to do with whether the game is running fullscreen or not. You can see that with it turned off, behaviour is still the same:

<details>
<summary>click</summary>

https://user-images.githubusercontent.com/56885706/154804209-3878a3b2-4c25-4ed7-b320-56df825a85ac.mp4
</details>


> Yep, it does sound like the user is running in exclusive fullscreen but wants to be running in borderless window. Note that this will increase input latency as a result.

I am well aware that the borderless mode exists on lazer, but just like on stable it is not an option for the reason you mentioned yourself (also performance is worse). This is not a solution.

**Reply by peppy** [reply]
stable there does not look to be running exclusive. how are you checking this?

that said, looking at your original video, it looks like a serious graphics driver bug. have you tried updating your drivers?

can you share your system specs?

**Reply by Joppe27** [reply]
> stable there does not look to be running exclusive. how are you checking this?

You are absolutely right. I wasn't verifying this because I assumed osu! would just be running fullscreen when I turned the option on. Stable's behaviour in true fullscreen as arguably worse: it has similar weird screen behaviour when alt-tabbing and sometimes even crashes explorer...

While taking another look, I found the solution for this issue.

The cause: Something to do with DPI scaling. Not sure how or what, I usually use 125% though on 100% scaling the issue still persists. I only know scaling is the issue because I happened to stumble onto the solution by accident.

The solution: When I override high DPI scaling behaviour and let "system" perform the scaling, my issue disappears. "System" is the default behaviour, so I have no idea how this would make any difference unless lazer does scaling itself and "system" overrides that. I am absolutely not informed on this, the only thing I know is that the issue is gone:
![afbeelding](https://user-images.githubusercontent.com/56885706/155545809-34c4008a-8ae7-4a16-80a9-74b600a8b594.png)

> that said, looking at your original video, it looks like a serious graphics driver bug. have you tried updating your drivers?

Yes, all of my system drivers are up-to-date including graphics drivers according to Windows Update and Lenovo Vantage.

> can you share your system specs?

[DxDiag.txt](https://github.com/ppy/osu/files/8133059/DxDiag.txt)
System: Thinkpad X1 Carbon (7th gen)
CPU: Intel i7-8565U CPU
GPU: Intel UHD Graphics 620
RAM: 16GB LPDDR3
OS: Windows 11 Pro, build 22000.527

**Reply by peppy** [reply]
Thanks, I think this can help us figure out the issue. It sounds like during the alt-tab process the window may be being resized multiple times due to the DPI changes.

If possible are you able to try the same thing with no external displays connected and see how it handles?

**Reply by Joppe27** [reply]
This is with the workaround, no external displays connected. As you can see, my "solution" does not seem to work without an external display:

https://user-images.githubusercontent.com/56885706/160644220-9d9a0569-0256-4fa8-b1b6-09a96c754c7d.mp4



If it helps, I also noticed that after alt-tabbing the osu! window is not in the correct position when previewing via the taskbar:

![PXL_20220329_150249959~2](https://user-images.githubusercontent.com/56885706/160644895-d824f45e-4a2d-4cd7-a31a-f355fde579c7.jpg)

**Comment by Susko3** [comment]
Try setting the resolution in osu! to the monitor native resolution (eg. change from `Default` to `1920x1080`) and see if it makes a difference.

If necessary, cycle the window modes (F11) and/or close and reopen the game.

**Reply by Susko3** [reply]
Also, could you try enabling GPU scaling in your GPU settings?

**Reply by Joppe27** [reply]
> Try setting the resolution in osu! to the monitor native resolution (eg. change from `Default` to `1920x1080`) and see if it makes a difference.
> 
> If necessary, cycle the window modes (F11) and/or close and reopen the game.

Sadly, this did not change the behavior.

> Also, could you try enabling GPU scaling in your GPU settings?

I don't think this setting exists for my Intel GPU. Are you using integrated graphics to test?


I also noticed that osu! does not manage to detect my second display's resolution correctly (both of my displays are 1920x1080):
![afbeelding](https://user-images.githubusercontent.com/56885706/163626036-b8c19aa5-5a07-4a85-b8fe-7da66000c96b.png)
For the record: I am using 125% display scaling on my primary display, and 100% on my second monitor.
1350p is exactly 125% of 1080p. Coincidence? 🤔
